// @chakra-ui/theme/dist/chakra-ui-theme.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var foundations_sizes_dist_chakraUiThemeFoundationsSizes = require('./sizes-dae9239f.cjs.dev.js');
var components_dist_chakraUiThemeComponents = require('../components/dist/chakra-ui-theme-components.cjs.dev.js');
var foundations_dist_chakraUiThemeFoundations = require('../foundations/dist/chakra-ui-theme-foundations.cjs.dev.js');
var utils = require('@chakra-ui/utils');
require('../foundations/spacing/dist/chakra-ui-theme-foundations-spacing.cjs.dev.js');
require('@chakra-ui/anatomy');
require('@chakra-ui/theme-tools');
require('../foundations/typography/dist/chakra-ui-theme-foundations-typography.cjs.dev.js');
require('@chakra-ui/system');
require('../foundations/borders/dist/chakra-ui-theme-foundations-borders.cjs.dev.js');
require('../foundations/breakpoints/dist/chakra-ui-theme-foundations-breakpoints.cjs.dev.js');
require('../foundations/colors/dist/chakra-ui-theme-foundations-colors.cjs.dev.js');
require('../foundations/radius/dist/chakra-ui-theme-foundations-radius.cjs.dev.js');
require('../foundations/shadows/dist/chakra-ui-theme-foundations-shadows.cjs.dev.js');
require('../foundations/transition/dist/chakra-ui-theme-foundations-transition.cjs.dev.js');
require('../foundations/z-index/dist/chakra-ui-theme-foundations-z-index.cjs.dev.js');
require('../foundations/blur/dist/chakra-ui-theme-foundations-blur.cjs.dev.js');

var semanticTokens = {
  colors: {
    "chakra-body-text": {
      _light: "gray.800",
      _dark: "whiteAlpha.900"
    },
    "chakra-body-bg": {
      _light: "white",
      _dark: "gray.800"
    },
    "chakra-border-color": {
      _light: "gray.200",
      _dark: "whiteAlpha.300"
    },
    "chakra-placeholder-color": {
      _light: "gray.500",
      _dark: "whiteAlpha.400"
    }
  }
};

var styles = {
  global: {
    body: {
      fontFamily: "body",
      color: "chakra-body-text",
      bg: "chakra-body-bg",
      transitionProperty: "background-color",
      transitionDuration: "normal",
      lineHeight: "base"
    },
    "*::placeholder": {
      color: "chakra-placeholder-color"
    },
    "*, *::before, &::after": {
      borderColor: "chakra-border-color",
      wordWrap: "break-word"
    }
  }
};
var styles$1 = styles;

var requiredChakraThemeKeys = ["borders", "breakpoints", "colors", "components", "config", "direction", "fonts", "fontSizes", "fontWeights", "letterSpacings", "lineHeights", "radii", "shadows", "sizes", "space", "styles", "transition", "zIndices"];
function isChakraTheme(unit) {
  if (!utils.isObject(unit)) {
    return false;
  }

  return requiredChakraThemeKeys.every(function (propertyName) {
    return Object.prototype.hasOwnProperty.call(unit, propertyName);
  });
}

var direction = "ltr";
var config = {
  useSystemColorMode: false,
  initialColorMode: "light",
  cssVarPrefix: "chakra"
};
var theme = foundations_sizes_dist_chakraUiThemeFoundationsSizes._extends({
  semanticTokens: semanticTokens,
  direction: direction
}, foundations_dist_chakraUiThemeFoundations["default"], {
  components: components_dist_chakraUiThemeComponents["default"],
  styles: styles$1,
  config: config
});

exports["default"] = theme;
exports.isChakraTheme = isChakraTheme;
exports.requiredChakraThemeKeys = requiredChakraThemeKeys;
exports.theme = theme;
